qiskit/primitives/decorators/__init__.py

```diff
@@ -0,0 +1,148 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Decorators for primitives
+=========================
+
+Allow Decorators
+================
+
+.. currentmodule:: qiskit.primitives.decorators.allow
+
+Allow decorators enhance Primitives (Sampler and Estimator) and provide a more flexible interface.
+For example, the default Primitives requires that ``circuit_indices`` cannot be optional or must
+be given a list of integers. Otherwise, it returns a type error as follows:
+
+.. jupyter-execute::
+   :hide-code:
+
+    from qiskit.primitives import (
+        Estimator,
+        Sampler,
+        allow_broadcasting,
+        allow_objects,
+        allow_optional,
+    )
+    from qiskit import QuantumCircuit
+    from qiskit.quantum_info import SparsePauliOp
+    from qiskit.circuit.library import RealAmplitudes
+
+    ansatz = RealAmplitudes(num_qubits=2, reps=2).decompose()
+    parameter = list(range(6))
+    circuit = QuantumCircuit(2)
+    circuit.x(1)
+    circuit.measure_all()
+    observable = 0.1 * SparsePauliOp("XX") + 0.4 * SparsePauliOp("ZZ")
+
+
+When an object is passed instead of an integer,
+
+.. jupyter-execute::
+   :raises:
+
+    with Estimator([ansatz], [observable]) as est:
+        result = est([ansatz], [observable], [parameter])
+    print(result)
+
+
+When nothing is passed,
+
+.. jupyter-execute::
+   :raises:
+
+    with Sampler([circuit]) as sampler:
+       result = sampler()
+    print(result)
+
+Allow Object
+------------
+
+:func:`~allow_objects` allows objects as inputs of indices instead of integer.
+For example, the following code works:
+
+.. jupyter-execute::
+
+    @allow_objects
+    class CustomEstimator(Estimator):
+        ...
+
+    with CustomEstimator([ansatz], [observable]) as e:
+        result = e(
+            circuit_indices=[ansatz],
+            observable_indices=[observable],
+            parameter_values=[parameter],
+        )
+    print(result)
+
+Object equivalency is determined by the object ID. Thus, note that even if they represent the same
+circuit (e.g. ``copy()``), they will not match if they are different objects.
+Also, note that since qiskit's circuit objects and observable objects are not immutable,
+if the object is changed after the construction of the primitives, it may behave unexpectedly.
```

Comment:
What happens when the object inside the "context" doesn't match with those in the `with ... as ...`? Since we are passing the objects, would it be possible to make it so that we only pass objects inside the "context" like:

```python
with CustomEstimator() as e: # no need to pass anything
        result = e(
            circuit_indices=[ansatz],
            observable_indices=[observable],
            parameter_values=[parameter],
        )
```

Reply:
This extension is not impossible if a circuit update feature is included, but the communication and serialization overhead would be significant.

Reply:
I don't quite understand the what you said, probably because I don't understand the whole python context concept. Do you know some resources to understand the context manager better? I am going to read [PEP343](https://peps.python.org/pep-0343/) where this was introduced.

Do you know any major software that uses the context manager? I would like to see more code examples. The only thing I know is for openning files.

Reply:
This is not a context issue, but simply because the cost of circuit serialization and communication is high, so it is faster to communicate a small object like int.

(Major software: for example, `requests` has `Session()` that can be used with `with`.)